libcxx/test/benchmarks/algorithms/nonmodifying/find_first_of.bench.cpp

```diff
@@ -0,0 +1,115 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+
+#include <algorithm>
+#include <cstddef>
+#include <deque>
+#include <initializer_list>
+#include <list>
+#include <string>
+#include <vector>
+
+#include <benchmark/benchmark.h>
+#include "../../GenerateInput.h"
+
+int main(int argc, char** argv) {
+  auto std_find_first_of = [](auto first1, auto last1, auto first2, auto last2) {
+    return std::find_first_of(first1, last1, first2, last2);
+  };
+
+  // Benchmark {std,ranges}::find_first_of where we have a hit at 25% of the haystack
+  // and at the end of the needle. This measures how quickly we're able to search inside
+  // the needle.
+  {
+    auto bm = []<class Container>(std::string name, auto find_first_of) {
+      benchmark::RegisterBenchmark(
+          name,
+          [find_first_of](auto& st) {
+            std::size_t const size = st.range(0);
+            using ValueType        = typename Container::value_type;
+            ValueType x            = Generate<ValueType>::random();
+            ValueType y            = random_different_from({x});
+            ValueType z            = random_different_from({x, y});
+            Container haystack(size, x);
+            Container needle(size, y);
```

Comment:
We probably want to use different haystack and needle container, since we can do optimizations if just the haystack fulfills specific properties.

Reply:
Thinking about it more, I'm actually not certain what optimizations we can do if only the haystack fulfills some properties. I think the properties of the needle are more interesting, but then the optimizations would matter most when the needle is fairly large, which may not be a common case. I'm tempted to punt on this in the initial version of these benchmarks and explore this when we get to optimizing the algorithm.